### PR TITLE
Check Slatepack file size before loading contents

### PR DESCRIPTION
**Pull request by unseddd**
Check Slatepack file size is within valid range before loading contents into memory.

### CODE REVIEW

**Comment by antiochp** [comment]
Not sure `max_block_size` is the right thing to base slate size limit off.
A slate will potentially have a max size transaction plus additional data. May be easier to simply hardcode/configure a max size directly in wallet itself.

**Comment by unseddd** [comment]
> Not sure max_block_size is the right thing to base slate size limit off.
A slate will potentially have a max size transaction plus additional data. May be easier to simply hardcode/configure a max size directly in wallet itself.

I had similar reservations about directly basing max Slatepack size on max block size, since the metadata in block is not one-to-one with the metadata in Slatepack. However, hardcoding max size in wallet isn't fool-proof either, and has overhead for any changes to max block size in the node (i.e. developers will have to remember to manually update max Slatepack size).

Ideally, the max Slatepack size would be: `MAX_BLOCK - BLOCK_HEADER + SLATEPACK_HEADER` (maybe with some extra space for whitespace + `>`). Though, with Slatepack files, there should be no whitespace, since the files are presumably generated by wallets (not copy-pasted). IIUC, that would encode max transaction size + Slatepack header size + acceptable whitespace size. WDYT?

**Comment by unseddd** [comment]
> Not sure max_block_size is the right thing to base slate size limit off.
A slate will potentially have a max size transaction plus additional data. May be easier to simply hardcode/configure a max size directly in wallet itself.

I had similar reservations about directly basing max Slatepack size on max block size, since the metadata in block is not one-to-one with the metadata in Slatepack. However, hardcoding max size in wallet isn't fool-proof either, and has overhead for any changes to max block size in the node (i.e. developers will have to remember to manually update max Slatepack size).

Ideally, the max Slatepack size would be: `MAX_BLOCK - BLOCK_HEADER + SLATEPACK_METADATA` (maybe with some extra space for whitespace + `>`). Though, with Slatepack files, there should be no whitespace, since the files are presumably generated by wallets (not copy-pasted). Slatepack metadata size might be hard to calculate precisely, since it would vary between binary and JSON encoded Slatepacks.

IIUC, the above formula would encode max transaction size + Slatepack header size + acceptable whitespace size. WDYT?

**Comment by unseddd** [comment]
> Not sure max_block_size is the right thing to base slate size limit off.
A slate will potentially have a max size transaction plus additional data. May be easier to simply hardcode/configure a max size directly in wallet itself.

I had similar reservations about directly basing max Slatepack size on max block size, since the metadata in block is not one-to-one with the metadata in Slatepack. However, hardcoding max size in wallet isn't fool-proof either, and has overhead for any changes to max block size in the node (i.e. developers will have to remember to manually update max Slatepack size).

Ideally, the max Slatepack size would be: `MAX_BLOCK - BLOCK_HEADER + SLATEPACK_METADATA` (maybe with some extra space for whitespace + `>`). Though, with Slatepack files, there should be no whitespace, since the files are presumably generated by wallets (not copy-pasted). Slatepack metadata size might be hard to calculate precisely, since it would vary between binary and JSON encoded Slatepacks.

IIUC, the above formula would encode max transaction size + Slatepack metadata size + acceptable whitespace size. WDYT?

**Comment by davidtavarez** [comment]
It would be nice to write some tests for this.

**Comment by unseddd** [comment]
Updated with new `grin_api::max_tx_size` function, and calculation for additional Slatepack metadata. @antiochp let me know what you think.

File: libwallet/src/slatepack/packer.rs [hunk 1]
```diff
@@ -48,11 +48,6 @@ impl<'a> Slatepacker<'a> {
 
 	/// return slatepack
 	pub fn deser_slatepack(&self, data: Vec<u8>, decrypt: bool) -> Result<Slatepack, Error> {
```
Comment (antiochp) [inline libwallet/src/slatepack/packer.rs:50]: Is this called from places other than `get_slatepack_file_contents` ?
We may want to keep the size validation here to handle other non-file based cases.

**Comment by unseddd** [comment]
Tests are failing to compile due to `grin_core::global::max_tx_weight` not being tagged. Tests pass local testing, with some workarounds for versioning, changes unrelated to this PR, etc.

**Comment by yeastplume** [comment]
> Tests are failing to compile due to `grin_core::global::max_tx_weight` not being tagged. Tests pass local testing, with some workarounds for versioning, changes unrelated to this PR, et

For dev on master just run `cargo update` to get the latest version of master from grin core, it'll make changes to Cargo.lock but that's fine at this stage of development

**Comment by unseddd** [comment]
> For dev on master just run cargo update to get the latest version of master from grin core, it'll make changes to Cargo.lock but that's fine at this stage of development

Build still fails due to traits not being in scope or implemented. Any tips on resolving failures like [here](https://dev.azure.com/mimblewimble/grin-wallet/_build/results?buildId=2247&view=logs&j=0dc20d02-474e-542d-9a6f-624de3d86d64&t=a547b719-1c0f-5f77-0843-37e7625bb141&l=703) and [here](https://dev.azure.com/mimblewimble/grin-wallet/_build/results?buildId=2247&view=logs&j=75e8fe06-bfd6-5564-cb5c-dff4fd81606c&t=85d1db4b-a01f-5226-0dbf-eb363124d748&l=840) (here or another PR)?

**Comment by yeastplume** [comment]
> > For dev on master just run cargo update to get the latest version of master from grin core, it'll make changes to Cargo.lock but that's fine at this stage of development
> 
> Build still fails due to traits not being in scope or implemented. Any tips on resolving failures like [here](https://dev.azure.com/mimblewimble/grin-wallet/_build/results?buildId=2247&view=logs&j=0dc20d02-474e-542d-9a6f-624de3d86d64&t=a547b719-1c0f-5f77-0843-37e7625bb141&l=703) and [here](https://dev.azure.com/mimblewimble/grin-wallet/_build/results?buildId=2247&view=logs&j=75e8fe06-bfd6-5564-cb5c-dff4fd81606c&t=85d1db4b-a01f-5226-0dbf-eb363124d748&l=840) (here or another PR)?

hrm.. crap. It looks as if some dependency is trying to pull in a different version of ed25519 dalek then we use. Let me look into this and fix via another PR.

**Comment by unseddd** [comment]
> Let me look into this and fix via another PR.

Sounds good to me, will rebase over the fix.

**Comment by yeastplume** [comment]
@unseddd Should be all good now, could you try again rebasing with latest master please?

**Comment by unseddd** [comment]
@yeastplume the fix worked, many thanks :smile:

**Pull request merged**